test/torchaudio_unittest/functional/functional_impl.py

```diff
@@ -45,6 +45,34 @@ def test_shape(self, shape):
         output_waveform = F.lfilter(waveform, a_coeffs, b_coeffs)
         assert shape == waveform.size() == output_waveform.size()
 
+    def test_9th_order_filter_stability(self):
+        """
+        Testing numerical error when using high order filter.
+
+        The filter coefficients `b` and `a` was get by `scipy.signal.butter(9, 850, 'hp', fs=22050, output='ba')`,
+        and the file `"22050_9th_butter_850_hp_ir_1024.pt"` storing the target impulse response of the filter which
+        was pre-computed using `scipy.signal.butter(9, 850, 'hp', fs=22050, output='sos')`.
+
+        The hi-pass frequency was selected so that when using double precision, the absolute error is less then 1e-4;
+        but when using single precision, the absolute error can be greater than 1.
+        """
+        ir_filepath = common_utils.get_asset_path('22050_9th_butter_850_hp_ir_1024.pt')
+        y = torch.load(ir_filepath, map_location=self.device)
+        y = y.to(self.dtype)
+        x = torch.zeros(1024, dtype=self.dtype, device=self.device)
+        x[0] = 1
+        a = torch.tensor([1., -7.60545606, 25.80187885, -51.23717435,
+                          65.62093428, -56.20096964, 32.18274279, -11.88025302,
+                          2.56506938, -0.24677075], dtype=self.dtype, device=self.device)
+        b = torch.tensor([0.49676025, -4.47084227, 17.88336908, -41.72786118,
+                          62.59179178, -62.59179178, 41.72786118, -17.88336908,
+                          4.47084227, -0.49676025], dtype=self.dtype, device=self.device)
+        yhat = F.lfilter(x, a, b, False)
+        if self.dtype == torch.float64:
+            self.assertEqual(yhat, y, atol=1e-4, rtol=1e-5)
+        else:
+            self.assertEqual(yhat, y, atol=10, rtol=1e-5)
```

Comment:
So my concerns about this approach were
1. it is not intuitively clear what the second assertion is intended.
2. if the implementation is updated and once `float32` passes the test, the code does not give the signal and we might forget to update the test by removing the branching here.

So, how about the following approach?
1. We keep the only the correct assertion.
```python
yhat = F.lfilter(x, a, b, False)
self.assertEqual(yhat, y, atol=1e-4, rtol=1e-5)
```
2. For `float32`, we overload the method in `functional_cpu_test.py` and decorate it with [`unittest.expectedFailure`](https://docs.python.org/3/library/unittest.html#unittest.expectedFailure) where `dtype=float32`.
```python
# in functional_cpu_test.py and functional_cuda_test.py
class TestLFilterFloat32(Lfilter, common_utils.PytorchTestCase):
    ...

    @unittest.expectedFailure
    def test_9th_order_filter_stability(self):
        super().test_9th_order_filter_stability()
```

This way, the test implementation looks like what is intended, and we can correctly acknowledge the failure for `float32`. If we update the implementation and the test starts to pass for `float32`, then the test will be reported as `unexpected success` so that we will not forget to remove the special treatment for `float32`.

Reply:
Awesome, let's do this.